arangod/Zkd/ZkdHelper.cpp

```diff
@@ -264,37 +272,91 @@ auto zkd::compareWithBox(byte_string_view cur, byte_string_view min, byte_string
       continue;
     }
 
-    if (!isLargerThanMin[dim]) {
+    if (!isLargerLowerThanMinMax[dim].first) {
       if (cur_bit == Bit::ZERO && min_bit == Bit::ONE) {
         result[dim].outStep = step;
         result[dim].flag = -1;
       } else if (cur_bit == Bit::ONE && min_bit == Bit::ZERO) {
-        isLargerThanMin[dim] = true;
+        isLargerLowerThanMinMax[dim].first = true;
         result[dim].saveMin = step;
       }
     }
 
-    if (!isLowerThanMax[dim]) {
+    if (!isLargerLowerThanMinMax[dim].second) {
       if (cur_bit == Bit::ONE && max_bit == Bit::ZERO) {
         result[dim].outStep = step;
         result[dim].flag = 1;
       } else if (cur_bit == Bit::ZERO && max_bit == Bit::ONE) {
-        isLowerThanMax[dim] = true;
+        isLargerLowerThanMinMax[dim].second = true;
         result[dim].saveMax = step;
       }
     }
+
+    dim += 1;
+    if (dim >= dimensions) {
+      dim = 0;
+      step += 1;
+    }
   }
 
-  return result;
 }
 
 auto zkd::testInBox(byte_string_view cur, byte_string_view min, byte_string_view max, std::size_t dimensions)
 -> bool {
-  auto cmp = compareWithBox(cur, min, max, dimensions);
 
-  return std::all_of(cmp.begin(), cmp.end(), [](auto const& r) {
-    return r.flag == 0;
-  });
+  if (dimensions == 0) {
+    auto msg = std::string{"dimensions argument to "};
+    msg += __func__;
+    msg += " must be greater than zero.";
+    throw std::invalid_argument{msg};
+  }
+
+  std::size_t max_size = std::max(std::max(cur.size(), min.size()), max.size());
+
+  BitReader cur_reader(cur);
+  BitReader min_reader(min);
+  BitReader max_reader(max);
+
+  ::arangodb::containers::SmallVector<std::pair<bool, bool>>::allocator_type::arena_type a;
+  ::arangodb::containers::SmallVector<std::pair<bool, bool>> isLargerLowerThanMinMax{a};
+
+  unsigned dim = 0;
+  unsigned finished_dims = 2 * dimensions;
+  for (std::size_t i = 0; i < 8 * max_size; i++) {
+
+    auto cur_bit = cur_reader.next().value_or(Bit::ZERO);
+    auto min_bit = min_reader.next().value_or(Bit::ZERO);
+    auto max_bit = max_reader.next().value_or(Bit::ZERO);
+
+    if (!isLargerLowerThanMinMax[dim].first) {
+      if (cur_bit == Bit::ZERO && min_bit == Bit::ONE) {
+        return false;
+      } else if (cur_bit == Bit::ONE && min_bit == Bit::ZERO) {
+        isLargerLowerThanMinMax[dim].first = true;
+        if (--finished_dims == 0) {
+          break;
+        }
+      }
+    }
+
+    if (!isLargerLowerThanMinMax[dim].second) {
+      if (cur_bit == Bit::ONE && max_bit == Bit::ZERO) {
+        return false;
+      } else if (cur_bit == Bit::ZERO && max_bit == Bit::ONE) {
+        isLargerLowerThanMinMax[dim].second = true;
+        if (--finished_dims == 0) {
+          break;
+        }
+      }
+    }
+
+    dim += 1;
+    if (dim >= dimensions) {
+      dim = 0;
+    }
```

Comment:
@maierlars is that better than `dim = dim % dimensions`?

Reply:
perf says so. This is actually branch free using `cmovxx`.

Reply:
Interesting, and unexpected.